subsys/zbus/proxy_agent/zbus_proxy_agent.c

```diff
@@ -0,0 +1,788 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/zbus/proxy_agent/zbus_proxy_agent.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(zbus_proxy_agent, CONFIG_ZBUS_PROXY_AGENT_LOG_LEVEL);
+
+/* Forward declaration */
+static int zbus_proxy_agent_send(struct zbus_proxy_agent_config *config,
+				 struct zbus_proxy_agent_msg *msg, uint8_t transmit_attempts);
+static int zbus_proxy_agent_stop_tracking(struct zbus_proxy_agent_config *config, uint32_t msg_id);
+static void zbus_proxy_agent_ack_timeout_handler(struct k_work *work);
+
+static bool is_duplicate_message(struct zbus_proxy_agent_config *config, uint32_t msg_id)
+{
+	if (!config) {
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < config->duplicate_detection.detection_buffer_size; i++) {
+		if (config->duplicate_detection.detection_buffer[i] == msg_id) {
+			LOG_DBG("Detected duplicate message ID %d", msg_id);
+			return true;
+		}
+	}
+	return false;
+}
+
+static void add_to_duplicate_detection(struct zbus_proxy_agent_config *config, uint32_t msg_id)
+{
+	if (!config) {
+		return;
+	}
+
+	config->duplicate_detection.detection_buffer[config->duplicate_detection.detection_head] =
+		msg_id;
+	config->duplicate_detection.detection_head =
+		(config->duplicate_detection.detection_head + 1) %
+		config->duplicate_detection.detection_buffer_size;
+
+	LOG_DBG("Added message ID %d to duplicate detection buffer at position %d", msg_id,
+		config->duplicate_detection.detection_head);
+}
+
+static uint32_t calculate_backoff_timeout(struct zbus_proxy_agent_config *config, uint8_t attempts)
+{
+	uint32_t multiplier;
+	uint32_t timeout;
+
+	multiplier = 1U << attempts;
+
+	/* Check for overflow */
+	if (attempts >= 32 || config->tracking.ack_timeout_start_ms > (UINT32_MAX >> attempts)) {
+		return config->tracking.ack_timeout_max_ms;
+	}
+
+	timeout = config->tracking.ack_timeout_start_ms * multiplier;
```

Comment:
You may use math extras to remove that check: https://docs.zephyrproject.org/latest/doxygen/html/group__math__extras.html